woocommerce/google-listings-and-ads
# Shipping Zone Integration - Refactor Shipping Rates API

### Changes proposed in this Pull Request:



Implements part of #1117.

This PR modifies the following APIs so that they can accept and return multiple shipping rates/methods per country (instead of just returning a single rate):

- `GET/POST mc/shipping/rates` 
- `POST mc/shipping/rates/batch`
- `GET mc/shipping/rates/suggestions`

With this change, the APIs now accept the following data schema:
```JSON
{
  "country_code": "US",
  "rates": [
    {
      "method": "flat_rate",
      "currency": "USD",
      "rate": 50,
      "options": {
        "free_shipping_threshold": "100"
      }
    }
  ]
}
```

The `method` parameter specifies the shipping method and for now, can only be set to `flat_rate`. However, once support for the pickup method is added, it can also be set to `local_pickup`.

The `options` parameter accepts an object that contains the `free_shipping_threshold` parameter. This option can be set for the `flat_rate` shipping method to specify the minimum order value for which the free shipping option should be offered.

To offer an unconditional free shipping method, the `rate` parameter can be set to `0`. The `free_shipping_threshold` option will NOT be considered if the shipping rate parameter is `0`. In other words, we can't set a condition if the shipping is already free for the country.

The `ShippingRateTable` now includes two new columns to store the shipping `method` and its `options` alongside each shipping rate. In order to make this change, I had to modify some of the existing code to be able to handle cases where multiple rows were returned for one country. As an example, the `Query` class now also accepts a `groupby` clause that allows creating grouped queries. Moreover, a new DB migration class is introduced that will update the DB once the plugin is updated.

I've also modified the country code validation method that was used in the shipping rates API and a few other places so that it can differentiate between a single value and an array. A new `get_country_codes_array_validate_callback` method is added that only works on arrays of country codes. This is unrelated to this feature but it was a needed improvement because I noticed that the API accepts array values where it should only accept one singular value.

Note that as part of the changes here, `offers_free_shipping` and `free_shipping_threshold` have been removed from the `/mc/settings` endpoint.

### Detailed test instructions:

**Test the DB migration:**

1. Update the `WC_GLA_VERSION` constant to a higher version than the current one.
2. Go to the `Migration20211228T1640692399::get_applicable_version` method and modify the `x.x.x` version to be equal to or greater than the one you set above.
3. Open any page in your website admin
4. Confirm that the `wp_gla_shipping_rates` table is updated and includes the `options` and `method` columns. The `method` column must be set by default to `flat_rate`


**Test the APIs:**

1. Define multiple shipping zones each containing various regions and shipping methods. Make sure you have at least one free shipping method with a minimum order amount option.
2. Make a `GET` request to the `mc/shipping/rates/suggestions` API with a `country_code` parameter and confirm the rate that is returned is the same that you defined in your shipping zone settings.
3. Make a `POST` request to the `mc/shipping/rates` containing the following JSON data:

```JSON
{
  "country_code": "DE",
  "rates": [
    {
      "method": "flat_rate",
      "currency": "USD",
      "rate": 50,
      "options": {
        "free_shipping_threshold": "100"
      }
    }
  ]
}
```

4. Make two `GET` requests to the `mc/shipping/rates/US` and `mc/shipping/rates` APIs and confirm the shipping rate you set above is returned correctly by these APIs.
5. Make a `POST` request to the `mc/shipping/rates/batch` API and send the following JSON data:

```JSON
{
  "country_codes": [
    "DE",
    "FR"
  ],
  "rates": [
    {
      "method": "flat_rate",
      "currency": "USD",
      "rate": 30,
      "options": {
        "free_shipping_threshold": "150"
      }
    }
  ]
}
```

6. Repeat step 4
7. Make a `POST` request to the `wc/gla/mc/settings/sync` API
8. Confirm that the shipping rates you have just set are reflected in your Merchant Center account (https://merchants.google.com/mc/shipping/services). Check the minimum order value as well.


### Changelog entry

> Add - Integration with WooCommerce Shipping Zone to automatically sync shipping settings to Merchant Center
